src/js/components/componentDiagram/graph/index.js

```diff
@@ -0,0 +1,308 @@
+
+        node.group = nodeGroup;
+        node.element = nodeGroup.element;
+
+        this.nodesGroup.appendChild(nodeGroup.element);
+      }
+    });
+
+    this.graph.edges().forEach(({ v, w }) => {
+      const edge = this.graph.edge(v, w);
+
+      if (edge.group) {
+        edge.group.move(edge.points);
+      } else {
+        const edgeGroup = new EdgeGroup(edge.points, this.options.animation);
+        edgeGroup.element.classList.add(`nodes---${v}---${w}`);
+
+        edge.group = edgeGroup;
+        edge.element = edgeGroup.element;
+
+        this.edgesGroup.appendChild(edgeGroup.element);
+      }
+    });
+
+    this.element.setAttribute('width', this.graph.graph().width);
+    this.element.setAttribute('height', this.graph.graph().height);
+  }
+
+  clearHighlights() {
+    this.outputGroup.querySelectorAll('.highlight,.highlight--inbound').forEach((el) => {
+      el.classList.remove('highlight');
+      el.classList.remove('highlight--inbound');
+    });
+  }
+
+  highlightNode(id) {
+    const highligthedNode = this.graph.node(id);
+    if (!highligthedNode || highligthedNode.element.classList.contains('dim')) {
+      return false;
+    }
+
+    highligthedNode.element.classList.add('highlight');
+
+    this.graph.nodeEdges(id).forEach((e) => {
+      const edge = this.graph.edge(e).element;
+      edge.classList.add('highlight');
+
+      if (id === e.w) {
+        edge.classList.add('highlight--inbound');
+      }
+
+      // Render highlighted connections above non-highlighted connections
+      if (!edge.classList.contains('dim')) {
+        const parent = edge.parentNode;
+        parent.removeChild(edge);
+        parent.appendChild(edge);
+      }
+    });
+
+    return true;
+  }
+
+  clearFocus() {
+    this.outputGroup.querySelectorAll('.dim').forEach((el) => {
+      el.classList.remove('dim');
+    });
+  }
+
+  focus(id) {
+    const [visitedNodes, visitedEdges] = findTraversableNodesAndEdges(this.graph, id);
+
+    this.graph.nodes().forEach((nodeId) => {
+      if (visitedNodes.has(nodeId)) {
+        return;
+      }
+
+      const node = this.graph.node(nodeId);
+      if (node.type !== 'cluster') {
+        node.element.classList.add('dim');
+      }
+    });
+
+    this.graph.edges().forEach((edgeId) => {
+      if (visitedEdges.has(edgeId)) {
+        return;
+      }
+
+      const edge = this.graph.edge(edgeId).element;
+      edge.classList.add('dim');
+
+      const parent = edge.parentNode;
+      parent.removeChild(edge);
+      parent.insertAdjacentElement('afterbegin', edge);
+    });
+  }
+
+  expand(id) {
+    this.graph.edges().forEach(({ v, w }) => {
+      if (v === id || w === id) {
+        const edge = this.graph.edge(v, w);
+        if (edge.group) {
+          edge.group.remove();
+          this.graph.removeEdge(v, w);
+        }
+      }
+    });
+
+    this.removeNode(id);
+
+    this.render();
+  }
+
+  collapse(pkg) {
+    this.render();
+  }
+
+  scrollToNodes(container, nodes) {
```

Comment:
UX note:

It was unclear to me why my viewport would move and zoom occasionally when expanding packages. With the movement of the viewport along with the movement of the nodes it's easy to lose track of what's going on. What if this method only translates the view and we drop the scaling?

Reply:
In some cases when user expands the package which contains many classes he might not understand what's going on. I've recorded 2 videos:

- First - with current behaviour
- Second - without zooming out when package expands

If you want not to zoom out then user expands the package - I'll disable zooming.

https://user-images.githubusercontent.com/17301016/103816546-e0263a80-50b0-11eb-871b-88239c48bee5.mov


https://user-images.githubusercontent.com/17301016/103816560-e4525800-50b0-11eb-9947-c0a9d838ba42.mov

Reply:
The first video you linked does look better. Maybe what I'm seeing is a bug. If you expand a package while zoomed out it sometimes zooms in strangely.
![Jan-07-2021 12-24-15](https://user-images.githubusercontent.com/8737782/103923659-5ba0df80-50e3-11eb-99bc-28d21199cb28.gif)